### ISSUE TITLE
[HOLD for payment 2022-12-20] [$8000] iOS Safari - Blue text selection persists in the compose box even after message is sent @metehanozyurt

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#8592]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
In iOS Settings, Go to General > Keyboard and enable autocorrection

1. Go to https://new.expensify.com on Safari iOS
2. Log in with any account and select any chat
3. Enter incorrectly spelled word in the compose box and let the autocorrection warn you with the blue text selection.
4. Send the message

## Expected Result:

Compose box should be cleared and empty blue selection shouldn't persist.

## Actual Result:

Blue text selection still exist.

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Mobile Web (safari)

**Version Number:** 1.1.92.0 (Matt A retested 8/26/22)

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** n/a

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/162746866-ab9925a0-99c9-4f8f-b63b-2c0601647649.mp4

https://user-images.githubusercontent.com/93399543/162746890-ad84201d-c88b-4bba-9ae4-7d2210b4f1ef.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** @metehanozyurt

**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1648452635209129

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MonilBhavsar** [comment]
Can't seem to reproduce on Safari web, but can reproduce on iOS simulator

**Comment by MonilBhavsar** [comment]
Asked for confirmation on a PR which I think caused this.

**Comment by parasharrajat** [comment]
I am not clear on the reproduction steps. 

> Write text to text field but incorrect word to input area warn you with blue area

What do we have to do here?

**Comment by parasharrajat** [comment]
Looked at the issue. Tried a few things to see if those fix it but no luck. I will try more tomorrow.

**Comment by MonilBhavsar** [comment]
Will get back to this

**Comment by MonilBhavsar** [comment]
@parasharrajat since this seems like a regression, do you want to give it a try to fix this?

**Comment by parasharrajat** [comment]
I would be happy to do that but I have been preoccupied with other stuff. I will try to debug this over the weekend and let you know by Monday.

**Comment by parasharrajat** [comment]
I tried many different approaches to solve the problem. But none of them worked. Basically, the selection length is already 0 so there is no reset. I tried to solve 7798 with a different approach than what I have on the PR but this issue persists.

I suggest exporting this if someone can fix this. My ultimate goal was to keep the spellcheck behavior and only try to reset the blue highlight.

**Comment by mallenexpensify** [comment]
@MonilBhavsar , any reason this shouldn't be labeled `External`?  
If it's confirmed to be a regression, let's add it to the [RCA list](https://docs.google.com/spreadsheets/d/1u73RE8J1DdX3mDohV1RXgulF6_cke3N_a15Sk-UUdQo/edit#gid=0) even if we don't know what the offending/root issue or PR is yet.

**Comment by MonilBhavsar** [comment]
No reason, adding external label now.
I'm not sure if we should add it to RCA list as it's not the breaking anything

**Comment by arielgreen** [comment]
https://www.upwork.com/jobs/~01208db9c91b51fa69

**Comment by arielgreen** [comment]
[Price increased](https://expensify.slack.com/archives/C01GTK53T8Q/p1652920061238279)

**Comment by songlang1994** [comment]
## Proposal
I believe it's an issue of Safari that we can fix it on the DOM side.

But if you do need to "fix" this issue, a workaround is disabling the `autocorrect` feature.

```diff
// File: src/components/Composer/index.js

<RNTextInput
  autoComplete="off"
+ autoCorrect={false}          // always disable auto correct
+ autoCorrect={isMobileSafari} // or more precisely
  placeholderTextColor={themeColors.placeholderText}
  ref={el => this.textInput = el}
```

**Comment by parasharrajat** [comment]
Thanks for the proposal. But as I said here https://github.com/Expensify/App/issues/8592#issuecomment-1116074045. The ultimate goal should be to fix the issue without disabling autoCorrect.

**Comment by songlang1994** [comment]
@parasharrajat I've read your comment before I post the proposal. But I misunderstood that you mean the `spellCheck` prop. Spell checking and auto correcting are two different features.

Let's wait a more appropriate solution.

**Comment by mananjadhav** [comment]
Awaiting proposals

**Comment by arielgreen** [comment]
Price increased

**Comment by allroundexperts** [comment]
@parasharrajat How about doing a `blur` on click of send button?

**Comment by parasharrajat** [comment]
We never give consider `setTimeout` hacks as solutions. But this can be used as last resort.

What is the effect of this change on all platforms? What will be your plan for all the platforms?

**Comment by parasharrajat** [comment]
Yeah, Focus should be remained on the composer after submit. That's what is the main motive.

**Comment by arielgreen** [comment]
Price increased

**Comment by dhairyasenjaliya** [comment]
https://github.com/Expensify/App/blob/11a1f0debac2dfee461e6b71751c3b3bcfb9f2cf/src/pages/home/report/ReportActionCompose.js#L411-L430

**Comment by dhairyasenjaliya** [comment]
### Proposal

- when we add any message which is a typo (misspelled) into the composer and send it as a message it sends as a message, but also it remains in text input with the auto-correct word in iOS (app) and for iOS (web safari) it removes the message, but blue opacity for spell check remain as it is 

### Solution

1. For Safari (web app) it's an issue of Safari that we can not fix on the DOM side.

2. To fix the issue in iOS (App) we need to comment out the line setNativeProps in submit() function as we are already clearing the text from the composer after submitting text so applying selection to empty text causes this issue 

    this.textInput.setNativeProps({selection: {start: 0, end: 0}});

https://github.com/Expensify/App/blob/11a1f0debac2dfee461e6b71751c3b3bcfb9f2cf/src/pages/home/report/ReportActionCompose.js#L425-L430

Like this : 

     this.props.onSubmit(trimmedComment);
     this.updateComment(''); // here we are already clearing composer textinput so applying 0 selection below this has no effect 
     this.setTextInputShouldClear(true);
     // Important to reset the selection on Submit action
     //this.textInput.setNativeProps({selection: {start: 0, end: 0}}); // comment this line solves problem in iOS (app)

**Note**: commenting on this line won't affect any other platform as we are already clearing the composer 

### Result (After commenting setNativeProps line) :  
Device : iPhone 13(App)

https://user-images.githubusercontent.com/47522946/173188941-a898f924-0cd1-455b-97f5-35d35ee38f0f.mp4

### Result (Without commenting setNativeProps line Current behaviour) :  

https://user-images.githubusercontent.com/47522946/173188991-d0c0ecf2-dcbc-4ea5-b165-69431e46cc8f.mp4

**Comment by arielgreen** [comment]
@mananjadhav

**Comment by MTN718** [comment]
# **Proposal**

Current issue from Safari mobile web browser. Because of auto correct text not initializing on Virtual Keyboard when update TextInput.

### Solution 

Add blur on Textinput when submit 
Add focus on After submit done

![image](https://user-images.githubusercontent.com/12084368/173641057-24f76935-3d8b-48e5-b5b2-a659355bd0e1.png)



https://user-images.githubusercontent.com/12084368/173642232-ffdd6586-aba2-4cd2-9d9e-acfaec81ffaa.mov

**Comment by mananjadhav** [comment]
@parasharrajat I don't have enough context on this and I can see you've been involved in this one. Would you mind taking up C+ on this one?

**Comment by parasharrajat** [comment]
Yeah, sure.

**Comment by mananjadhav** [comment]
@arielgreen Can you please assign @parasharrajat as C+ on this one?

**Comment by MTN718** [comment]
Any feedback for my proposal?

**Comment by parasharrajat** [comment]
@MTN718 this a duplicate proposal.

**Comment by MTN718** [comment]
So it's not solution for this issue?

**Comment by MonilBhavsar** [comment]
Assigned Rajat, oh no this is wrong. Please ignore

**Comment by MTN718** [comment]
This is not available to post proposal?

**Comment by MonilBhavsar** [comment]
No sorry, my bad. This is open for proposals.
But as @parasharrajat mentioned, your proposal seems similar to the one [we already received](https://github.com/Expensify/App/issues/8592#issuecomment-1140283813) and we don't want to go with that.

**Comment by MTN718** [comment]
Ok But I have tried lots of ways to resolve this issue. But that's the only way to resolve this issue I have tried. :)

**Comment by parasharrajat** [comment]
> For Safari (web app) it's an issue of Safari that we can not fix on the DOM side (blue background).

@dhairyasenjaliya This issue is mainly focused on fixing the issue on iOS web.

**Comment by parasharrajat** [comment]
ok, Please report it on Slack then. But it is very likely that both of those will be fixed with single solution.

**Comment by aneequeahmad** [comment]
@parasharrajat is this bug about the blue highlight(`blue border-color`) on text input after the message is sent ? I still dont understand the bug. It is on all platforms. Please help me understand the bug.

**Comment by parasharrajat** [comment]
It must be clear from the above discussion. But yeah the issue is that the blue highlighted color is left on the composer after the message is sent. 

This affects Mweb. There is another issue with iOS _but also it remains in text input with the auto-correct word in iOS (app)_ https://github.com/Expensify/App/issues/8592#issuecomment-1152926071

**Comment by MonilBhavsar** [comment]
Making it Monthly so we cap the price of this issue
https://expensify.slack.com/archives/C01SKUP7QR0/p1653671315618679
@arielgreen

**Comment by dhairyasenjaliya** [comment]
### Proposal 

**Root Cause**

- hello since I was debugging this issue I found out that this issue is mainly caused because we have added the mouseDown event ( e.preventDefault() ) in send button  in order to keep the focus on text input even after the message is sent, preventDefault method will stop all the default event for text input and keeps keyboard open, although we do clear the text values from text input when spellchecker is enabled iOS keeps the text values in suggestion and simple .clear() is not being able to clear from root (suggestion) for iOS, in order to reset the values along with suggestion we need to force update text input in order to actual reset text input with the suggestion also keeping the focus on  

**mouseDown Event** 

https://github.com/Expensify/App/blob/84f6bbe72702621e85f172a3e2e168728ab4f1e5/src/pages/home/report/ReportActionCompose.js#L656

<img width="856" alt="Screenshot 2022-06-28 at 9 17 48 PM" src="https://user-images.githubusercontent.com/47522946/176223605-908ac711-59fd-4275-b6b8-92f7ba61c5f8.png">

### Solution 

**Safari Mobile**

- In order to solve this issue we need to add a key property to the composer for safari and update this key only when we press send button this will force update the text input with reset text + clear suggestions from previously sent message and will remove blue opacity as expected while keeping focus and keyboard open 

**The following changes: only in index.js (web)**
 https://github.com/Expensify/App/blob/84f6bbe72702621e85f172a3e2e168728ab4f1e5/src/components/Composer/index.js#L123-130
 
 From 
 
    this.state = {
            numberOfLines: 1,
            selection: {
                start: initialValue.length,
                end: initialValue.length,
            },
        };
        
  To
  
    this.state = {
            numberOfLines: 1,
            selection: {
                start: initialValue.length,
                end: initialValue.length,
            },
            keyUpdate: false,
        };
        
        
  https://github.com/Expensify/App/blob/84f6bbe72702621e85f172a3e2e168728ab4f1e5/src/components/Composer/index.js#L161
  
  
 From 
 
        if (!prevProps.shouldClear && this.props.shouldClear) {
            this.textInput.clear();
            // eslint-disable-next-line react/no-did-update-set-state
            this.setState({numberOfLines: 1});
            this.props.onClear();
        }
        
   To
   
           if (!prevProps.shouldClear && this.props.shouldClear) {
            // eslint-disable-next-line react/no-did-update-set-state
            this.setState({
                numberOfLines: 1,
                // eslint-disable-next-line react/no-access-state-in-setstate
                keyUpdate: !this.state.keyUpdate,    // key is updated only when we textinput has shouldClear true
            }, () => {
                this.textInput.clear();
                this.props.onClear();
            });
        } 
       
        
 https://github.com/Expensify/App/blob/84f6bbe72702621e85f172a3e2e168728ab4f1e5/src/components/Composer/index.js#L350
       
 From 

       <RNTextInput
                autoComplete="off"
                placeholderTextColor={themeColors.placeholderText}
                ref={el => this.textInput = el}
                selection={this.state.selection}
                onChange={() => {
                    this.updateNumberOfLines();
                }}
                onSelectionChange={this.onSelectionChange}
                numberOfLines={this.state.numberOfLines}
                style={propStyles}
                /* eslint-disable-next-line react/jsx-props-no-spreading */
                {...propsWithoutStyles}
                disabled={this.props.isDisabled}
            />
            
            

 To
        
        <RNTextInput
                key={this.state.keyUpdate} // key updated only when textinput has shouldClear true
                autoComplete="off"
                placeholderTextColor={themeColors.placeholderText}
                ref={el => this.textInput = el}
                selection={this.state.selection}
                onChange={() => {
                    this.updateNumberOfLines();
                }}
                onSelectionChange={this.onSelectionChange}
                numberOfLines={this.state.numberOfLines}
                style={propStyles}
                /* eslint-disable-next-line react/jsx-props-no-spreading */
                {...propsWithoutStyles}
                disabled={this.props.isDisabled}
            />

 
 ### Result
 
 

https://user-images.githubusercontent.com/47522946/176236276-38d6ae0e-1aac-4ff4-a46d-1ab093c8b34f.mp4


 
 
 ### Without changes
 

https://user-images.githubusercontent.com/47522946/176235809-2022458a-e79d-43f0-ad4e-ee1c1883c2ae.mp4




**iOS app**

- To fix the iOS app I suggest my old proposal https://github.com/Expensify/App/issues/8592#issuecomment-1152926071

**Comment by dhairyasenjaliya** [comment]
yup agreed have tried a lot of different ways to solve this issue, and this solution only fits best in given conditions from my side, sure let's wait for better ideas

**Comment by arielgreen** [comment]
Reposted, the Upwork job automatically closed. https://www.upwork.com/jobs/~0141f966437b7764e6

**Comment by dhairyasenjaliya** [comment]
@parasharrajat how about we add a fix with this proposal https://github.com/Expensify/App/issues/8592#issuecomment-1152926071 which will solve 2 issues 
1. blue opacity for iOS (app) (for safari its issue for long time and yet we have no fix from apple this is only reference I can find https://stackoverflow.com/questions/18689842/clearing-input-element-in-mobile-safari-with-javascript-does-not-clear-the-ios-a)

2. If we add a typo in text input it will send a message and remain in text input with auto correct value https://github.com/Expensify/App/issues/10148

**Comment by parasharrajat** [comment]
I will discuss this @dhairyasenjaliya.

**Comment by parasharrajat** [comment]
Asked here https://expensify.slack.com/archives/C02NK2DQWUX/p1659432359765889

**Comment by parasharrajat** [comment]
Ok, based on the discussion, we decided to break this into two issues. One for iOS and the second for mWeb Safari. Further details will be coming shortly...

**Comment by dhairyasenjaliya** [comment]
alright, will wait for additional details :)

**Comment by arielgreen** [comment]
@parasharrajat per [this comment](https://github.com/Expensify/App/issues/8592#issuecomment-1203057240), what's the status here?

**Comment by MonilBhavsar** [comment]
@arielgreen, we're still discussing the issue

**Comment by mallenexpensify** [comment]
Per an internal discussion about monthly labels, bumped back to weekly so we can try to get this fixed soon. 

@MonilBhavsar are next best steps to 
1. Update the OP to make it clear this issue is specific to fix mWeb (I'm assuming we'll leave the $4000 price then double price a week after if we don't have any takers
2. Create a new issue/job focused on iOS.  If so, what should the starting price be for that job @MonilBhavsar, @parasharrajat

**Comment by mallenexpensify** [comment]
Thanks @parasharrajat, I'll wait for @MonilBhavsar to chime in about next steps too, I'm unsure what we can/should do here.

**Comment by dhairyasenjaliya** [comment]
Am I eligible to get compensation here? As I have added an initial proposal for solving the iOS blue opacity here https://github.com/Expensify/App/issues/8592#issuecomment-1152926071 which eventually gets resolved recently into my another PR which is to solve auto-correct word iOS app issue as mentioned here https://github.com/Expensify/App/issues/8592#issuecomment-1200671587  

- As this issue not only focuses on safari mWeb here at the start issue is updated with the iOS app as well https://github.com/Expensify/App/issues/8592#issuecomment-1105019343  

- I have also added more insight and found the exact root cause here (for safari) https://github.com/Expensify/App/issues/8592#issuecomment-1168985203

**Comment by mallenexpensify** [comment]
I'm still a bit confused here so bear with me....

1. @dhairyasenjaliya you fixed https://github.com/Expensify/App/issues/10291 for iOS native app with this PR - https://github.com/Expensify/App/pull/10457 so you'll be compensated for that at the amount stated $250. 
2. This issue (title and OP) has been updated to specifically address iOS Safari.  @dhairyasenjaliya , do you anticipate the fix for #10291 will automatically fix this issue as well?
3. If no to #2 you will need to get an approved proposal from @parasharrajat and @MonilBhavsar for this issue then a PR submitted, approved then deployed to production.  If that all works, then you'll be compensated $4k for this issue a week after it's been deployed to production, assuming there are no regressions

**Comment by dhairyasenjaliya** [comment]
hm @mallenexpensify actually when I started working on this issue, it was affecting both the iOS app and safari then I added a proposal for the iOS app-specific fix here on 11th June https://github.com/Expensify/App/issues/8592#issuecomment-1152926071 but somehow it gets to hold for internal discussion and then one more issue for auto-correct text got issued in git with the same proposal it gets fixed which I posted on 11th June we have added in the latest PR on 19th august


**So here my initial proposal got over-ridden with the new issue which I have fixed eventually for this issue**

**Comment by dhairyasenjaliya** [comment]
@parasharrajat here as we can see the title is updated to iOS on 21st April (for both app and mobile safari) https://github.com/Expensify/App/issues/8592#issuecomment-1105019343, on 16th June you mentioned issue only specific to mWeb https://github.com/Expensify/App/issues/8592#issuecomment-1157915558 perhaps you might miss out the issue was already updated for both platform

**Comment by arielgreen** [comment]
https://expensify.slack.com/archives/C02NK2DQWUX/p1662463071985089?thread_ts=1659432359.765889&cid=C02NK2DQWUX

**Comment by arielgreen** [comment]
@dhairyasenjaliya can you confirm whether you will be submitting a separate proposal for this issue?

**Comment by dhairyasenjaliya** [comment]
@arielgreen currently I have no separate proposal for this issue, I have added a solution just for the iOS app and did try to solve it for safari as well but no luck, so currently this is open to all contributors

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by arielgreen** [comment]
@mvtglobally https://github.com/Expensify/App/issues/8592#issuecomment-1259529887

**Comment by arielgreen** [comment]
Retested this on Safari mWeb, iOS 16.0 -- no longer reproducible.

**Comment by MonilBhavsar** [comment]
I'm able to reproduce this. @arielgreen Did you have "autocorrection" setting toggled on?

**Comment by arielgreen** [comment]
@MonilBhavsar yes I do. Are you on iOS 16?

**Comment by arielgreen** [comment]
Waiting on confirmation

**Comment by MonilBhavsar** [comment]
Yes, I'm on iOS 16 too

**Comment by MonilBhavsar** [comment]
https://user-images.githubusercontent.com/32012005/198282013-86ade9a6-d615-48cf-8b0d-b3845682cee2.MP4

**Comment by arielgreen** [comment]
Thanks Monil. Confirmed I can reproduce too.

**Comment by arielgreen** [comment]
Reposted here with an increased price: https://www.upwork.com/jobs/~01a221ac893d2d0e6f

**Comment by Dev-Abdul-Aleem** [comment]
It seems like this is either related to state management or input field value or css, I just want you guide me about file directory and then I will give you a detailed brief.

**Comment by MonilBhavsar** [comment]
@Dev-Abdul-Aleem you can find the code around compose box in this file https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose.js
You can find more instruction about app's directory structure and architecture here - https://github.com/Expensify/App#app-structure-and-conventions
If you have any questions, feel free to ask in #expensify-open-source slack channel.

**Comment by arielgreen** [comment]
@MonilBhavsar heads up we are now keeping E/App issues as daily

**Comment by MonilBhavsar** [comment]
We didn't had proposals to review so bumped to Weekly.

**Comment by MonilBhavsar** [comment]
Still waiting for proposals...

**Comment by 24dev** [comment]
### **Proposal**: 

This is actually a longstanding WebKit _feature_. It's caused by the autoCorrect buffer not being cleared on a `value` update like it would be updated when firing a keyboard event, and according to Darin Adler (engineering manager at Safari/WebKit) this was done on purpose, so may explain why it has never changed. You can read some of the reports [here](https://bugs.webkit.org/show_bug.cgi?id=139473). Even massive engineering teams that do support chat on mobile Safari, like LinkedIn, retain this same behaviour to this day (though obviously most chat apps just force you to use their app, to avoid these kinds of inconsistencies). However, since your budget for this has continued to grow, clearly changing this is important to you, so I have gone ahead with a proposal.

So, we know the root of this behaviour is that clearing an input value !== pressing backspace on the final character (what would normally clear the memory) - however we cannot hijack the backspace function, for obvious reasons, so that rules a direct fix out. We can try and fire `execCommand` directly to bypass this feature, and I did try this, but it is aggressively [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand) and felt pretty hacky so I doubted you guys would go for it, even though technically this would qualify as the most direct fix. However, it's there to consider. But rather than brute-forcing the native functionality to make it do what you want, I would suggest you step around it with the lightest footprint possible. 

I read through all your other proposals that focused on remounting, e.preventDefault(), blur/focus, and I agree none are ideal, but they are the right way of thinking about this, even if for the wrong reasons. Effectively, somehow we need to force autoCorrect to reset its memory without actually having access to autoCorrect memory directly (and we also can't toggle autoCorrect on/off for the input without closing the keyboard). 

So after a lot of thought, in my opinion, the the best fix for this is to use a honeypot input (like one would have to prevent bots). A hidden input, which we `.focus()` on after sending a message, then refocus our real input immediately after. This forces the autoCorrect buffer for the main input to reset, while keeping the keyboard active the whole time because we never `.blur()` anything. It is not expensive like remounting, does not open doors to future browser incompatibility, has no UX issues like other options, does not rely on the dreaded setTimeout, removes the need for your `preventDefault()` call on the submit button (a nice bonus), and only requires a few extra lines of code to your Composer component. It could even be useful as an actual honeypot input too - but I don't know enough about your product to say for sure! Additionally, if WebKit ever decide to change this bug/feature, this will be backwards compatible. Most importantly though, I believe the tradeoff here vs the ugly floating blue highlight is justified where it may not have been with previous options. So I've attached a video demonstrating it.

I haven't spent enough time working with you guys to know your exact codebase preferences (this will be my first bounty for you), so cannot comment on whether this is perfect for your codebase or not. This is what I would do if it were one of my apps and I wanted this behaviour changed. But either way, you should now have all the information you need here to close this issue and implement a solution. I can open a PR with this, or with an alternate workaround of your choosing, just let me know. In the meantime, I will begin hunting through the rest of your bounties. 

https://user-images.githubusercontent.com/17879993/200148653-06aa0233-f22c-46b6-b449-559fb0d7ec5f.mov

**Comment by parasharrajat** [comment]
Thanks for the proposal @24dev. Do you mind adding some code samples to show us how would you implement both `execCommand` and honeypot in our codebase?  

we have discussed honeypot in a few issues in the past. But decided to not go with these workarounds. I will reconsider this technique for this issue if sounded like the best fit.

Links: 
- [CONTRIBUTING Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md)
- [App overview and local development](https://github.com/Expensify/App#local-development)
- [CODE OF CONDUCT](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md)
- [Code style and patterns](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md)
- [Styling Components](https://github.com/Expensify/App/blob/main/contributingGuides/STYLING.md)

**Comment by arielgreen** [comment]
@24dev can you confirm whether you're able to add some code samples as Rajat mentioned above?

**Comment by 24dev** [comment]
@arielgreen Will have them up this evening!

**Comment by 24dev** [comment]
@arielgreen @parasharrajat 

### Honeypot

This integrates nicely with your existing `Composer` component and requires very little code. Using your browser utils from `src/libs/Browser/index.web.js` to determine browser. 

![Untitled design-2](https://user-images.githubusercontent.com/17879993/200654923-cb4acf7f-91a0-47be-9d78-8299f102da21.png)

---

### Exec Command
I tried `execCommand` and it looks like it doesn't work as a fix anymore, at least not for my Safari version. It's possible it could work on other Safari versions so I've shared the primary options you can try if you're really keen on it. I also tried all the other possible commands that might work but no luck. Like I mentioned before, it's possible they changed this in the years since it was discussed. But to reiterate, it's so heavily deprecated that even if it does work for you, it might not work in a few months/years. I noticed you are using it in other parts of the app too - something to keep a close eye on. 
<img width="576" alt="Screenshot 2022-11-08 at 17 26 23" src="https://user-images.githubusercontent.com/17879993/200651792-6a7a133f-6974-49be-b384-30055134c88b.png">

I also found this in an old WebKit patch report, but for me using a space doesn't help.
<img width="559" alt="Screenshot 2022-11-08 at 17 28 35" src="https://user-images.githubusercontent.com/17879993/200634220-068d042d-fad5-4a87-989d-623788b50b5a.png">

For confirmation that this is the reason for the issue, I also tried a number of other DOM editing methods like `insertNode`, `innerHTML`, `innerText`, `value`, `selection.modify` but none work either, which makes sense.

**Comment by arielgreen** [comment]
@parasharrajat update please

**Comment by parasharrajat** [comment]
I am reviewing it. I will post an update tomorrow.

**Comment by parasharrajat** [comment]
Putting up on my top priority. Expect an update tomorrow morning CST.

**Comment by 24dev** [comment]
@parasharrajat Let me know if you want me to do some digging too, as it's still fresh in my mind. Just tell me what direction you want to move in.

**Comment by JmillsExpensify** [comment]
@MonilBhavsar To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by parasharrajat** [comment]
So it seems that `execCommand` is not working. 

For HoneyPot, I raised a discussion https://expensify.slack.com/archives/C01GTK53T8Q/p1668726058235669. This is a hack so the majority needs to agree before we think about it.

**Comment by parasharrajat** [comment]
Discussing on slack.

**Comment by 24dev** [comment]
@parasharrajat Let me know if you think Honeypot won't be approved and I can look into some other options for us, but if the reception has been positive I can help on other bugs instead.

**Comment by parasharrajat** [comment]
Yeah, please share alternatives if they are not hacks/workarounds. 

You are free to send proposals to other issues as well. Only the condition is to work on one at a time for new contributors.

**Comment by parasharrajat** [comment]
@24dev Do let us know if you don't find any alternatives.

**Comment by dylanexpensify** [comment]
@parasharrajat any updates here?

**Comment by parasharrajat** [comment]
Waiting on https://github.com/Expensify/App/issues/8592#issuecomment-1322553692 and the internal discussion to decide the next steps.

**Comment by dhairyasenjaliya** [comment]
@parasharrajat I see now we have options only for a workaround so do you think adding a key to the composer (only to safari) will work now as I proposed previously here at that time we were not encouraging a workaround 

- IMO focusing & unfocus on a fake text input feels a bit hacky 

Previous proposal: https://github.com/Expensify/App/issues/8592#issuecomment-1168985203

**Result**

https://user-images.githubusercontent.com/47522946/203465172-9783f04f-443f-4257-aadb-bbf581eff4e4.mp4

**Comment by parasharrajat** [comment]
Remounting is very expensive. But let me check that as well.

**Comment by tjferriss** [comment]
@MonilBhavsar I'm following up on [Jason's comment](https://github.com/Expensify/App/issues/8592#issuecomment-1319227829). It seems like we should take this issue internal at this point given it's age.

**Comment by parasharrajat** [comment]
Checking this now.

**Comment by parasharrajat** [comment]
@dhairyasenjaliya Do you think setting the key as a boolean value is a good option?

**Comment by dhairyasenjaliya** [comment]
@parasharrajat AfAI boolean will work well since component get updates only when we change state (keyUpdate state) value , so we just need to update state with different values that can be anything unique than previous

**Comment by s77rt** [comment]
It's better to use a string, so just do `.toString()`

**Comment by JmillsExpensify** [comment]
@MonilBhavsar @tjferriss Escalated this in chat, as we need to act with urgency and act on next steps, whether that's approving an proposal and getting that implemented, or whether that's taking this internal.

**Comment by s77rt** [comment]
This seems more like a bug in WebKit, any proposed solution must be a hack.
It's safer to disable autocorrect on Safari than implementing a hacky solution

**Comment by JmillsExpensify** [comment]
That sounds like a good solution. I agree that if it's in WebKit, then the core issue is upstream and we shouldn't go down the rabbit hole of implementing a fix for what isn't a "bug" in our codebase.

**Comment by dhairyasenjaliya** [comment]
I think we should keep the same behavior as native iOS and use key in order to solve this issue and also key is more common technique in react to solve this kind of issue

**Comment by s77rt** [comment]
@dhairyasenjaliya re-rending is not really a good solution, it may give the user the slow-app experience

**Comment by dhairyasenjaliya** [comment]
@s77rt just try solution i posted above i don't really find any slowness in app after implying

**Comment by JmillsExpensify** [comment]
In either case, let's have @MonilBhavsar and @parasharrajat weigh in on next steps. Can you both prioritize this come Monday? Thank you!

**Comment by parasharrajat** [comment]
It will be finalized on Monday.

**Comment by parasharrajat** [comment]
Ok, I can see new alternatives being proposed above. Let me check on those.

**Comment by parasharrajat** [comment]
I am discussing something with @MonilBhavsar and it might take a day for him to reply as he might not be online now. We will take action as soon as the discussion is over.

**Comment by parasharrajat** [comment]
We have discussed this through and considered various factors including all the proposals on this issue. Soon @MonilBhavsar will share the updates here.

**Comment by JmillsExpensify** [comment]
Ok thanks, let's get this one moving! It's one of the 11 oldest issues in the repo.

**Comment by 24dev** [comment]
So since disabling autocorrect obviously would not have been accepted as a proposal prior to my explanation (but I can see why you'd prefer that to a workaround), is there any plan to reward externals here? I read your guidelines so I understand rewards are at your discretion, but I did explain the root cause of this issue and gave you the knowledge needed to make your decision, and without that proposal I'm quite certain this issue would still be wide open and require a lot more internal time (money) spent on investigation. In fact I think the first proposal on this issue _was_ to disable autoCorrect and it was rejected, making this even more of a grey area. Keen to hear your thoughts..

**Comment by parasharrajat** [comment]
> but I did explain the root cause of this issue and gave you the knowledge needed to make your decision, and without that proposal I'm quite certain this issue would still be wide open and require a lot more internal time (money) spent on investigation.

I can understand that time was invested in finding the root cause. But I disagree to the point that we took an action after your suggestions. We were already familiar with the limitations and this was investigated a long back over two issues. 

I agree that you proposed good workarounds however, we kept this issue open to find a solution. I have taken this internally to think about your proposal but as a team, we decided not to implement these workarounds for the low-value issues. 

Due to our internal priorities, we are clearing the backlog of very old issues. Whether there are proposals or not, we would have marked it internal. This is not only the issue where such actions are taken. A notification was issued [two weeks back](https://github.com/Expensify/App/issues/8592#issuecomment-1319227829).   

>  In fact I think the first proposal on this issue was to disable autoCorrect and it was rejected, making this even more of a grey area

This is being discussed and we will update if needed.

> is there any plan to reward externals here?

This is at our discretion and is being discussed as I said above. 


Please feel free to look into new issues or active ones. Looking forward to your contributions.

**Comment by JmillsExpensify** [comment]
Linked PR is merged.

**Comment by mallenexpensify** [comment]
Hi all, this is a 🐯 of an issue (over 200 comments) and we need to suss out payments. 
1. @parasharrajat are you able to help? 
2. If so... have you been assigned since the beginning, looks like it? (asking cuz, if you've invested time all along, include reviewing the PR, I imagine your compensation rate would be higher.) 
3. If yes to the above, and you've invested a considerable amount of time, is there any reason you **wouldn't/shouldn't** be due the original $8000 job price? 
4. Here's where things might get tricky.   We made this issue internal, @MonilBhavsar and @parasharrajat , did we use a substantial amount of code from any previously submitted proposals?  If so, can you link to them, tag the contributor and guesstimate how much of their code you used? (in increments of 25%, 50% or ~100%).  
5. Once we have details from #4 we'll work on potential compensation amounts.

**Comment by mallenexpensify** [comment]
> have you been assigned since the beginning, looks like it? (asking cuz, if you've invested time all along, including reviewing the PR, I imagine your compensation rate would be higher.)

@parasharrajat, what I meant here is, assuming we have the option for payments of $1000k (standard), 100%, 50% and 25% of the job amount.  I'm suggesting that your pay would be in the _higher_ bracket.    

Let me know when you and @MonilBhavsar reach consensus on if any code was used.  Thx

**Comment by mallenexpensify** [comment]
Thanks @parasharrajat 
Can I get one... last... final confirmation from @MonilBhavsar and @parasharrajat before I pay?  
- @parasharrajat $8000 for proposals and PR review
- @songlang1994 $2000 for 25% compensation for code submitted that ended up being used. 

Sound good?!?!?  
Please apply and I'll get paid once we confirm 
https://www.upwork.com/jobs/~019b27cbf20756346d

**Comment by mallenexpensify** [comment]
@MonilBhavsar ping for confirmation on the above when you're back.  Thx

**Comment by MonilBhavsar** [comment]
[This](https://github.com/Expensify/App/issues/8592#issuecomment-1368051227) plan looks good to me 👍

**Comment by arielgreen** [comment]
@metehanozyurt yes; can you please apply to the [job](https://www.upwork.com/jobs/~019b27cbf20756346d)?

**Comment by arielgreen** [comment]
@parasharrajat payment issued.

**Comment by metehanozyurt** [comment]
I applied, thank you @arielgreen .

**Comment by arielgreen** [comment]
Also sent offer to @songlang1994

**Comment by songlang1994** [comment]
Offer accepted. Thank you @arielgreen

**Comment by arielgreen** [comment]
Excellent, all parties have been paid.

**Comment by arielgreen** [comment]
@parasharrajat @MonilBhavsar can you please review the [checklist](https://github.com/Expensify/App/issues/8592#issuecomment-1349851877)?

**Comment by MonilBhavsar** [comment]
Done!

**Comment by arielgreen** [comment]
Thank you! Reviewing adding a regression test to this, then we'll be set to close it out.

**Comment by arielgreen** [comment]
Proposed test [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1674079621821409).
